---
title: "Differential_transcrip_expr"
author: "Brian Gudenas"
date: "August 18, 2017"
output: html_document
fig_width: 8
fig_height: 6
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, message = FALSE, cache.lazy = FALSE, warning = FALSE)
opts_knit$set(root.dir = "C:/Users/Brian/Documents/Localization")
```


```{r DL_Libraries, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("rhdf5")
devtools::install_github("pachterlab/sleuth")
```

```{r Libraries}
library(readr)
library(stringr)
library(sleuth)
library(dplyr)
library(ggplot2)
```


```{r Metadata}
## load Metadata and convert Salmon output for use in Sleuth using wasabi::prepare_fish_for_sleuth
samples = read_tsv("./Data/Meta/metadata.tsv")
colnames(samples) = make.names(colnames(samples))
Pair = read.table("./Data/Meta/Pair_map.txt", sep="\t", header = TRUE)

samples = samples[!is.na(match(samples$File.accession, Pair$Pair1)), ]

samples = samples[samples$Biosample.subcellular.fraction.term.name == "nucleus" | samples$Biosample.subcellular.fraction.term.name == "cytosol", ]

files = file.path(getwd(), "/Data/Results", samples$File.accession)


```

```{r Sleuth, cache = TRUE}

samples$Library.depleted.in[is.na(samples$Library.depleted.in)] = "rRNA"
samples$Lib_method = paste0(samples$Library.made.from, samples$Library.depleted.in)
s2c = dplyr::select(samples, sample = File.accession, condition = Biosample.subcellular.fraction.term.name, cell = Biosample.term.name, platform = Platform, Lib = Lib_method, reads = Read.length, warning = Audit.WARNING)

s2c = dplyr::mutate(s2c, path = files)
so = sleuth_prep(s2c, transformation_function = function(x) log2(x+0.5))
lds = plot_pc_variance(so)
plot_pca(so, color_by = "Lib",  text_labels = TRUE) +ggtitle("Total") + theme(text = element_text(size=20)) + xlab(paste("PC1","%Var =", round(lds$data[1,2],2))) + ylab(paste("PC2","Var =", round(lds$data[2,2],2)))
```

There appears to be large amount of variance in the Minus Poly A samples while samples from both other Lib types cluster very tightly. Therefore we will perform DTE within each lib type seperately.
```{r split, cache = TRUE }

RNA = dplyr::filter(s2c, Lib == "RNArRNA")
polyA = dplyr::filter(s2c, Lib == "polyadenylated mRNArRNA")
nopolyA = dplyr::filter(s2c, Lib == "RNArRNA, polyadenylated mRNA")


so_RNA = sleuth_prep(RNA, extra_bootstrap_summary = TRUE, transformation_function = function(x) log2(x + 0.5))
so_polyA = sleuth_prep(polyA, extra_bootstrap_summary = TRUE, transformation_function = function(x) log2(x + 0.5))
so_nopolyA = sleuth_prep(nopolyA, extra_bootstrap_summary = TRUE , transformation_function = function(x) log2(x + 0.5))
save.image("./Data/DTE.RData")

```

```{r DTE}
lds = plot_pc_variance(so_nopolyA)
plot_pca(so_nopolyA, color_by = "condition",  text_labels = TRUE) +ggtitle("PCA - Minus Poly A") + theme(text = element_text(size=20)) + xlab(paste("PC1","%Var =", round(lds$data[1,2],2))) + ylab(paste("PC2","Var =", round(lds$data[2,2],2)))

lds = plot_pc_variance(so_RNA)
plot_pca(so_RNA, color_by = "condition", text_labels = TRUE) +ggtitle("PCA - Total RNA") + theme(text = element_text(size=20)) + xlab(paste("PC1","%Var =", round(lds$data[1,2],2))) + ylab(paste("PC2","Var =", round(lds$data[2,2],2)))

lds = plot_pc_variance(so_polyA)
plot_pca(so_polyA, color_by = "condition",  text_labels = TRUE) +ggtitle("PCA - Poly A") + theme(text = element_text(size=20)) + xlab(paste("PC1","%Var =", round(lds$data[1,2],2))) + ylab(paste("PC2","Var =", round(lds$data[2,2],2)))
```

Appears to be two possible outliers in the Poly A samples (ENCFF901DUO, ENCFF792JEE) driving large variance in PC1. 
Upon inspection these samples contain audit warnings for "low read depth" so we will remove them 

```{r}
dplyr::filter(polyA, sample == "ENCFF901DUO" | sample == "ENCFF792JEE")
```


```{r, New_polyA, cache=TRUE}
polyA = dplyr::filter(polyA, sample != "ENCFF901DUO" & sample != "ENCFF792JEE")
so_polyA = sleuth_prep(polyA, extra_bootstrap_summary = TRUE,  transformation_function = function(x) log2(x + 0.5))

lds = plot_pc_variance(so_polyA)
plot_pca(so_polyA, color_by = "condition",  text_labels = TRUE) +ggtitle("PCA - Poly A") + theme(text = element_text(size=20)) + xlab(paste("PC1","%Var =", round(lds$data[1,2],2))) + ylab(paste("PC2","Var =", round(lds$data[2,2],2)))

nRNA = nrow(RNA)
npolyA = nrow(polyA)
nnopolyA = nrow(nopolyA)

save.image("./Data/DTE.RData")

```

Now we will perform differential expression analysis for each RNA-seq dataset. Multiple testing adjustments are already performed within each DE test therefore we will only adjust the overall FDR-threshold since 3 tests  are being performed. Therefore we selected the FDR threshold of 0.01 (0.01 < (0.05/3) )

```{r Model_nopolyA}

so_nopolyA <- sleuth_fit(so_nopolyA, ~cell + condition, 'full')
so_nopolyA <- sleuth_fit(so_nopolyA, ~cell, 'reduced')
so_nopolyA <- sleuth_lrt(so_nopolyA, 'reduced', 'full')
so_nopolyA = sleuth_wt(so_nopolyA, "conditionnucleus")
wt_results = sleuth_results(so_nopolyA, 'conditionnucleus')
lrt_results = sleuth_results(so_nopolyA, "reduced:full", "lrt")
res = merge(lrt_results, wt_results[ , c("target_id", "b", "se_b", "mean_obs")], on = "target_id", sort = FALSE)
res$target_id = unlist(lapply(str_split(res$target_id, "\\."), "[[", 1))
out_name = paste0("./Data/tables/DTE_nopolyA","_","condition_nucleus", ".tsv")
res = res[!is.na(res$qval), ]
res = res[res$qval <= 0.01, ]
write.table(res, out_name, sep = "\t", quote = FALSE, row.names = FALSE)

```

```{r Model_RNA}

so_RNA <- sleuth_fit(so_RNA, ~cell + condition, 'full')
so_RNA <- sleuth_fit(so_RNA, ~cell, 'reduced')
so_RNA <- sleuth_lrt(so_RNA, 'reduced', 'full')
so_RNA = sleuth_wt(so_RNA, "conditionnucleus")
wt_results = sleuth_results(so_RNA, 'conditionnucleus')
lrt_results = sleuth_results(so_RNA, "reduced:full", "lrt")
res = merge(lrt_results, wt_results[ , c("target_id", "b", "se_b", "mean_obs")], on = "target_id", sort = FALSE)
res$target_id = unlist(lapply(str_split(res$target_id, "\\."), "[[", 1))
out_name = paste0("./Data/tables/DTE_RNA","_","condition_nucleus", ".tsv")
res = res[!is.na(res$qval), ]
res = res[res$qval <= 0.01, ]
write.table(res, out_name, sep = "\t", quote = FALSE, row.names = FALSE)

```

```{r Model_polyA}
# add in platform covariate since polyA samples were run on genome analyzer or hiseq
so_polyA <- sleuth_fit(so_polyA, ~cell + platform + condition, 'full')
so_polyA <- sleuth_fit(so_polyA, ~cell + platform, 'reduced')
so_polyA <- sleuth_lrt(so_polyA, 'reduced', 'full')
so_polyA = sleuth_wt(so_polyA, "conditionnucleus")
wt_results = sleuth_results(so_polyA, 'conditionnucleus')
lrt_results = sleuth_results(so_polyA, "reduced:full", "lrt")
res = merge(lrt_results, wt_results[ , c("target_id", "b", "se_b", "mean_obs")], on = "target_id", sort = FALSE)
res$target_id = unlist(lapply(str_split(res$target_id, "\\."), "[[", 1))
out_name = paste0("./Data/tables/DTE_polyA","_","condition_nucleus", ".tsv")
res = res[!is.na(res$qval), ]
res = res[res$qval <= 0.01, ]
write.table(res, out_name, sep = "\t", quote = FALSE, row.names = FALSE)

```

```{r Merge_results}
  polyA = read_tsv("./Data/tables/DTE_polyA_condition_nucleus.tsv") 
      
  nopolyA = read_tsv("./Data/tables/DTE_nopolyA_condition_nucleus.tsv") 
     
  RNA = read_tsv("./Data/tables/DTE_RNA_condition_nucleus.tsv") 
         
  
  df = as.data.frame(rbind(cbind(polyA$target_id, polyA$b, "polyA"),
            cbind(nopolyA$target_id, nopolyA$b, "nopolyA"),
            cbind(RNA$target_id, RNA$b, "RNA")))
  colnames(df) = c("target_id","b","type")
  df$b = as.numeric(as.character(df$b))
  df$b = round(df$b, 3)
  
  nTotal =  94 # total number of samples
  #  set weights for each dataset to integrate the log fold-change estimates from the 3 DE tests
  df$wt = npolyA
  df$wt[df$type == "nopolyA"] = nnopolyA
  df$wt[df$type == "RNA"] = nRNA
  df$target_id = as.character(df$target_id)
  
  
  joined_df = matrix(nrow=length(unique(df$target_id)), ncol = 1)
  rownames(joined_df) = unique(df$target_id)
  colnames(joined_df) = "b"
  for (i in rownames(joined_df)) {
    trans = df[df$target_id == i, ]
    wt_sum = sum(trans$b * trans$wt)/sum(trans$wt)
    joined_df[ rownames(joined_df) == i, 1] = wt_sum
  }
  
joined_df = as.data.frame(joined_df)
```

```{r, LncRNA_test, eval = FALSE}
mart = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
map = getBM(mart = mart, attributes = c("ensembl_transcript_id","transcript_biotype", "transcript_length","percentage_gene_gc_content", "transcript_count"),
            filters = "ensembl_transcript_id", values=rownames(joined_df))


lncRNA_filter = c("3prime_overlapping_ncrna", "antisense","antisense_RNA", "lincRNA","ncrna_host","processed_transcript", "sense_intronic" , "sense_overlapping", "bidirectional_lncRNA", "non_coding", "macro_lncRNA")

lnc_test=c()
for (biotype in map$transcript_biotype) {
    lnc_test =c(lnc_test, (sum(grepl(biotype, lncRNA_filter)) > 0 ) )
}
mat = match(rownames(joined_df), map$ensembl_transcript_id)


joined_df$lncRNA = lnc_test[mat]
joined_df$transcript_biotype = map$transcript_biotype[mat]
joined_df$transcript_length = map$transcript_length[mat]
joined_df$percentage_gene_gc_content = map$percentage_gene_gc_content[mat]

joined_df = joined_df[!is.na(joined_df$lncRNA), ]
pc = joined_df[joined_df$transcript_biotype == "protein_coding", ]
lncRNAs = joined_df[joined_df$lncRNA == TRUE, ]

pdf("./Figures/LncRNA_PC_density.pdf")
ggplot(data= lncRNAs) +
  geom_density(data = pc, aes(x = b), col = "blue", lwd = 1.5) + 
  geom_density(aes(x = b), col = "red", lwd = 2) +
  theme_bw() + 
  theme(axis.text=element_text(size=14), axis.title=element_text(size=16,face="bold")) +
  ylab("Density") +
  xlim(c(-7,7)) +
  xlab(expression("Log"[2]* " Fold Change") ) +
  guides()
  ggl


## Check number of nuclear and cytosolic lncRNAs
library(gridExtra)
lncnum = data.frame("LncRNAs" = nrow(lncRNAs), Nuclear = sum(lncRNAs$b > 3), Cytosol = sum(lncRNAs$b < -1), row.names = NULL)
grid.table(lncnum, vp = grid::viewport(x = 0.4, y = 0.9), rows = NULL, theme = ttheme_default(base_size = 16))
dev.off()

seq = biomaRt::getSequence(id = rownames(lncRNAs), type="ensembl_transcript_id", seqType = "cdna", mart = mart)
lncRNAs$cdna = seq$cdna[match(rownames(lncRNAs), seq$ensembl_transcript_id)]
lncRNAs = lncRNAs[, -2]

saveRDS(lncRNAs, "./Data/DE_lncRNAs.rds")
```

```{r, genome_lncRNAs, cache = TRUE}
library(seqinr)
library(stringr)
library(biomaRt)

ncRNA = read.fasta("./Data/Homo_sapiens.GRCh38.ncrna.fa")
Annot = getAnnot.list(ncRNA)

  lncRNA_filter = c("3prime_overlapping_ncrna", "antisense","antisense_RNA", "lincRNA","ncrna_host","processed_transcript", "sense_intronic" , "sense_overlapping", "bidirectional_lncRNA", "non_coding", "macro_lncRNA")

lnc_test=c()
bios =c()

for (i in Annot){
  ID = strsplit(i[1][[1]], split = "\\.")[[1]][1]
  ID = str_sub(ID, 2, nchar(ID))
  biotype = strsplit(i[1][[1]], split = "transcript_biotype:")[[1]][2]
  biotype = strsplit(biotype, "gene_symbol:")[[1]][1]
  biotype = str_trim(biotype)
  
  if ( sum(grepl(biotype, lncRNA_filter)) > 0 ){ lnc_test =c(lnc_test, ID) 
                                            bios =c(bios, biotype)}
}


mart = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
map = getBM(mart = mart, attributes = c("ensembl_transcript_id","transcript_length","percentage_gene_gc_content", "transcript_count"), filters = "ensembl_transcript_id", values = lnc_test)
names(bios) = lnc_test
map$transcript_biotype = bios[match(map$ensembl_transcript_id, names(bios))]

seq = biomaRt::getSequence(id = map$ensembl_transcript_id, type="ensembl_transcript_id", seqType = "cdna", mart = mart)

map$cdna = seq$cdna[match(map$ensembl_transcript_id, seq$ensembl_transcript_id)]
rownames(map) = map$ensembl_transcript_id
map = map[ ,-1]
saveRDS(map, "./Data/genomic_lncRNAs.rds")

```

```{r Kmer_add, eval = FALSE}
# this function takes a df of transcripts (containing a "cdna" column of the sequence) and integer of K for which it calculates
# k-mer frequency features from 2:K and appends these columns to df
set.seed(54321)

lncRNAs = readRDS("./Data/DE_lncRNAs.rds")
lncRNAs = lncRNAs[lncRNAs$transcript_length >= 200, ]
inTrain = sample(1:nrow(lncRNAs), .7*nrow(lncRNAs) )

genome = readRDS("./Data/genomic_lncRNAs.rds")

Kmer_add = function(df, k){
  library(Biostrings)
  #df must contain "cdna" column
  #k is largest int of 
  kmer_mat = matrix( ncol = sum(4^seq(2, k)) , nrow = nrow(df), data =0)
  
  for (i in 1:nrow(df)){
    seq =DNAString(df$cdna[i])
   # len = df$transcript_length[i]
    index = 0
    for (j in 2:k) {
    kmers = oligonucleotideFrequency(seq, j)
    ind2 = index
    kmer_mat[i, (index + 1): (ind2 + (4^j) ) ] = as.vector(kmers)
    index = index + (4^j)
    }
  }
  
  nams =c()
  for (i in 2:k){
    nams =c(nams, names(oligonucleotideFrequency(seq, i)) )
  
  }
  colnames(kmer_mat) = nams
  df = cbind(df, kmer_mat)
  return(df)
}
############################################

Kmer_filt = function(df, topN){
#takes df with cols[6:ncol(df)] as kmer-counts and ranks kmers based on p-value from Spearman Correlation with L2FC selecting 1:topN of kmers
  library(broom)

  test_vals =c()


  for ( i in 6:ncol(df)) {
    vals = cor(df[ , i], df$b, method = "spearman")
    # vals = cor(df[ , i], df$b, method = "spearman")
    test_vals = c(test_vals, abs(vals ))
  }
  names(test_vals) = colnames(df[6:ncol(df)])

  #test_vals = p.adjust(test_vals, "fdr")
  test_vals = sort(test_vals, decreasing = TRUE)

  kmer_mat = match(names(test_vals), colnames(df[ ,6:ncol(df)]) )

  filt = kmer_mat[1:topN]

  filt = c(1:5, filt+5)

  return(filt)

}


Supervised = Kmer_add(lncRNAs, k = 5)
#filt = Kmer_filt(Supervised[inTrain, ], 5000)
#Supervised = Supervised[ , filt]

#Supervised[ ,6:ncol(Supervised)] = Supervised[ ,6:ncol(Supervised)]/Supervised$transcript_length
saveRDS(Supervised, "./Data/Supervised_kmer.rds")

genome = genome[genome$transcript_length >= 200, ]
Unsupervised = Kmer_add(genome, k = 5)
#Unsupervised = Unsupervised[ , filt]
#Unsupervised[ ,6:ncol(Unsupervised)] = Unsupervised[ ,6:ncol(Unsupervised)]/Unsupervised$transcript_length
saveRDS(Unsupervised, "./Data/Unsupervised_kmer.rds")
```

```{r, RNA_motifs, eval = FALSE}

RNABP_Motif_add = function(df){
  library(Biostrings)
  
motifs_dir = file.path("./Data/motifs/Homo_sapiens_2017_09_07_CISBP/pwms_all_motifs")
mot_num = length(list.files(motifs_dir))

mat_mot = matrix(nrow = nrow(df), ncol = mot_num , data = 0)

for (i in 1:mot_num){
    fils = list.files(motifs_dir)[i]
    #some motif files from CISBP are completely empty and will throw an error so i added a try-catch here
    mot = try(read.table(paste0(motifs_dir, "/", fils), sep = "\t", row.names = 1, header = TRUE ), silent = FALSE)
    if (class(mot) != "try-error"){
        mot = t(mot)
# rownames(mot)
# [1] "A" "C" "G" "U"
## Motifs are in units of RNA so we must complement to DNA (A -> T, C -> G, G -> C, U -> A)
        rownames(mot) = c("T","G","C", "A")
        mot = mot[order(rownames(mot)), ]
        #print(ncol(mot))
        
        for (j in 1:nrow(df)){
        #len = df$transcript_length[j]
        seq = df$cdna[j]
        counts = countPWM(mot, seq, min.score = "60%") 
        mat_mot[j,i] = counts
        }
    }    
}

df = cbind(df, mat_mot)
return(df)
}
Supervised = RNABP_Motif_add(Supervised)
drop_cols = which(colSums(Supervised[ ,6:ncol(Supervised)]) == 0)
Supervised = Supervised[ ,- match(names(drop_cols), colnames(Supervised))]

saveRDS(Supervised, "./Data/Supervised_train.rds")

Unsupervised = RNABP_Motif_add(Unsupervised)
Unsupervised = Unsupervised[ ,- match(names(drop_cols), colnames(Unsupervised))]
saveRDS(Unsupervised, "./Data/Unsupervised_train.rds")

```


```{r, eval = FALSE}

set.seed(54321)

Supervised = Supervised[ , -c(5 )]
#Supervised$b = round(Supervised$b, 2)

Supervised$transcript_biotype = as.factor(Supervised$transcript_biotype)
#Supervised = Supervised[Supervised$gene_biotype != "non_coding" & Supervised$gene_biotype != "macro_lncRNA", ]
#Supervised$gene_biotype = droplevels(Supervised$gene_biotype)
# Supervised$Localized = as.factor(Supervised$Localized)

inTrain = sample(1:nrow(Supervised), .7*nrow(Supervised) )
train = Supervised[inTrain, ]
test = Supervised[-inTrain, ]

#library(caret)
library(randomForest)


rf = randomForest(x = train[ , -1],
           y = train$b,
           #trControl = control,
           #tuneGrid = tunegrid,
           ntree = 501)
print(rf)
plot(rf)
#save.image("./Data/Raw_RF.RData")
saveRDS(rf, "./Data/Raw_RF.rds")

Loc =rep("Nuclear", nrow(train) )
Loc[train$b < 0] = "Cytosol"
Loc = as.factor(Loc)

rf2 = randomForest(x= train[ ,-1], 
                   y = Loc,
                   ntree = 501)



pred = predic

test_vals = predict(rf, test)
cor.test(test_vals, test$b)
sqrt(mean((test_vals - test$b)^2))

```





