---
title: Prediction of LncRNA Subcellular Localization with Deep Learning from Sequence Features
author: "Brian Gudenas"
date: "February 16, 2018"
output: word_document
---
# Prediction of LncRNA Subcellular Localization with Deep Learning from Sequence Features


```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, message = FALSE, cache = TRUE)
opts_knit$set(root.dir = "..")
```

```{r Libraries, warning = FALSE}
library(readr)
library(stringr)
library(sleuth)
library(dplyr)
library(RUVSeq)
library(RColorBrewer)
library(pheatmap)
```

## Load Metadata

```{r metadata,  warning = FALSE}
samples = read_tsv("./Data/Meta/metadata.tsv")
colnames(samples) = make.names(colnames(samples))
Pair = read.table("./Data/Meta/Pair_map.txt", sep="\t", header = TRUE)

samples = samples[!is.na(match(samples$File.accession, Pair$Pair1)), ]

samples = samples[samples$Biosample.subcellular.fraction.term.name == "nucleus" | samples$Biosample.subcellular.fraction.term.name == "cytosol", ]

samples$Library.depleted.in[is.na(samples$Library.depleted.in)] = "rRNA"
samples$Lib_method = paste0(samples$Library.made.from,"_", samples$Library.depleted.in)

s2c = dplyr::select(samples, sample = File.accession, condition = Biosample.subcellular.fraction.term.name, cell = Biosample.term.name, platform = Platform, Lib = Lib_method, reads = Read.length, warning = Audit.WARNING)

## Remove any cell types with less than 4 total samples
s2c = s2c[is.na(match(s2c$cell, names(table(s2c$cell))[table(s2c$cell) < 4] ) ),   ]

filespath = file.path(getwd(), "/Data/Kallisto/Quant", s2c$sample)
s2c = dplyr::mutate(s2c, path = filespath)

## Total samples
nrow(s2c)
## Samples per fraction
table(s2c$condition)
## Samples per Library type
table(s2c$Lib)
## Cell types
table(s2c$cell)

## Kallisto.o* is the standard output from the kallisto quant call for all samples
output = read.delim(file = "./Data/Kallisto/Kallisto.o2139569", sep = "\n")
pairmap = read.delim(file ="./Data/Meta/Pair_map.txt")

alignments = output[grepl("processed", output[,1]), ]

pairmap$total_reads = as.numeric(str_trim(str_replace_all(unlist(lapply(str_split(unlist(lapply(str_split(unlist(lapply(str_split(alignments, ", "), "[[", 1)), " reads"), "[[", 1)), "processed"),"[[",2)),",","")))
pairmap$reads_aligned = as.numeric(str_replace_all(unlist(lapply(str_split(unlist(lapply(str_split(alignments, ", "), "[[", 2)), " reads"), "[[", 1)), ",",""))

pairmap$percent_aligned = pairmap$reads_aligned/pairmap$total_reads*100
s2c$total_reads = pairmap$total_reads[match(s2c$sample, pairmap$Pair1)]
s2c$reads_aligned = pairmap$reads_aligned[match(s2c$sample, pairmap$Pair1)]

## total aligned reads
sum(s2c$reads_aligned)

rm(alignments, samples, pairmap, Pair, output)
write.csv(s2c, "./Data/Meta/Sample_Meta.csv")
```

## DTE by cell type
Perform differential transcript expression analysis for each cell type. If a cell type contains more than a single RNA library construction protocol then we utilize library type as a covariate in the DTE test.


```{r DTE_loop,  warning = FALSE}
master =c()
for ( i in unique(s2c$cell)){
  s2c_tmp = dplyr::filter(s2c, cell == i) %>% arrange(sample)
  
  so = sleuth_prep(s2c_tmp, transformation_function = function(x) log2(x+0.5), extra_bootstrap_summary = TRUE)
  lds = plot_pc_variance(so)

plot_pca(so, color_by = "condition",  text_labels = TRUE) + xlab(paste("PC1","%Var =", round(lds$data[1,2],2))) + ylab(paste("PC2","Var =", round(lds$data[2,2],2)))
  
## if more than 1  RNA-seq library protocols were used account for Lib covariate
if (length(unique(s2c_tmp$Lib))  > 1 ) {
    
    so <- sleuth_fit(so, ~ Lib + condition, 'full')
    so <- sleuth_fit(so, ~ Lib, 'reduced')
    so = sleuth_wt(so, "conditionnucleus")
    wt_results = sleuth_results(so, 'conditionnucleus')
    res =wt_results[!is.na(wt_results$pval), ]
    res$target_id = unlist(lapply(str_split(res$target_id, "\\."), "[[", 1))
    out_name = paste0("./Data/tables/DTE","_",i, ".tsv")
    write.table(res, out_name, sep = "\t", quote = FALSE, row.names = FALSE)
    res$cell = i
    
    master = rbind(master,res)
  
  } else {
    
     Expr = so$obs_norm %>%  dplyr::select(target_id, sample, est_counts)
    Emat = tidyr::spread(Expr, key  = sample, est_counts)
    rownames(Emat) = unlist(lapply(str_split(Emat[,1],"\\|"), "[[",1))
    Emat = as.matrix(Emat[,-1])
    #dim(Emat)
    ## need to round est_counts to integer for RUVseq
    Emat = round(Emat)
    
    # Filter transcripts 
    filter <- apply(Emat, 1, function(x) length(x[x > 5]) >= 2) ##remove non-expressed transcripts (5 counts     in 2 samples)
    filtered <- Emat[filter,]   
    EDASeq::plotPCA(filtered, col=as.numeric(as.factor(s2c_tmp$condition))+2, cex=1.2, k=3, pch = as.numeric(as.factor(s2c_tmp$Lib)), main = paste("Sleuth",i) )
    
    so <- sleuth_fit(so, ~  condition, 'full')
    so <- sleuth_fit(so, ~ 1, 'reduced')
    so <- sleuth_lrt(so, 'reduced', 'full')
    so = sleuth_wt(so, "conditionnucleus")
    wt_results = sleuth_results(so, 'conditionnucleus')
    res =wt_results[!is.na(wt_results$pval), ]
    res$target_id = unlist(lapply(str_split(res$target_id, "\\."), "[[", 1))
    out_name = paste0("./Data/tables/DTE","_",i, ".tsv")
    write.table(res, out_name, sep = "\t", quote = FALSE, row.names = FALSE)
    res$cell = i

    master = rbind(master,res)
  }

}

write.csv(master, "./Data/tables/Master.csv", quote = FALSE, row.names = FALSE)
rm(Emat,Expr,filtered,Pair,pairmap, res, samples, wt_results, alignments, filter, set,set1,so, output, s2c_tmp, filespath, i, out_name, wts,x)
```

## LncRNA Filter

```{r, lncRNA_preprocess}
master = read.csv("./Data/tables/Master.csv", header = TRUE, stringsAsFactors = FALSE)

annots = rtracklayer::readGFF(gzfile("./Data/Annotation/gencode.v28.long_noncoding_RNAs.gtf.gz"), version = 2L)
annots = dplyr::filter(annots, type == "transcript")
# remove trailing decimal
annots$transcript_id = lapply(strsplit(annots$transcript_id, "\\."), "[[", 1)

lnc_match = match(master$target_id, annots$transcript_id)

lncRNAs = master[!is.na(lnc_match), ]

## set up Expression matrix lncRNAs x cell_type
mat = matrix(nrow = length(unique(lncRNAs$target_id)), ncol = length(unique(lncRNAs$cell)), data = NA)
colnames(mat) = unique(lncRNAs$cell)
rownames(mat) = unique(lncRNAs$target_id)

## Loop through and add fold-changes, leave NA if not detected
for (i in 1:nrow(lncRNAs)){
  l2fc = lncRNAs$b[i]
  target = lncRNAs$target_id[i]
  celltype = lncRNAs$cell[i]
  
  mat[rownames(mat) == target, colnames(mat) == celltype] = l2fc
}

wts = table(s2c$cell)/nrow(s2c)
wts = wts[order(names(wts)) ]

mat = mat[ , order(colnames(mat))]

df = data.frame(target_id = rownames(mat))
df$l2fc = NA

for (i in 1:nrow(mat)){
  
  df$l2fc[i] = weighted.mean(mat[i,], wts, na.rm = TRUE)
}

```

## Feature Extraction

```{r, feature_extraction}
source("./src/DeepLncRNA.R")
mart = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
trans = GetSeq(df$target_id, mart)
trans = cbind(df$l2fc[match(trans$ensembl_transcript_id, df$target_id)], trans)
colnames(trans)[1] = "l2fc"
df = FeatureExtract(trans)

```

## Histogram of LncRNA localization and class selection
```{r, localization_histogram}
quantile(df$l2fc)
## Select top quantile as positive set and bottom quantile as negative set
## Round to nice numbers (Cytosolic = < 0, Nuclear >= 2.8)
# Make Histogram of L2FC dist
h = hist(df$l2fc ) 
cuts <- cut(h$breaks, c(-Inf,-1,2.8,Inf))

plot(h, col = c("blue","white","red")[cuts], xlim = range(-7, 7), xlab = expression("Log"[2]* " Fold-Change"), main = "", cex.lab = 1.2)
abline(v = median(df$l2fc), col = "grey", lwd = 2)

cyto = sum(df$l2fc < 0)
nuc = sum(df$l2fc > 2.8)

text(paste(cyto, "Cytosolic"), x = -2, y = 2200, col = "blue", cex = 1.3)
text(paste(nuc, "Nuclear"), x = 5.8, y = 2200, col = "red", cex = 1.3)

```


## Create training dataset

```{r partition_frames}
rownames(df) = df$ensembl_transcript_id
df = df %>% dplyr::select( -c(cdna, ensembl_transcript_id))

drop_vars = colnames(df[ ,4:ncol(df)])[colSums(df[ ,4:ncol(df)])==0]
df = df %>% dplyr::select( -one_of(drop_vars))
saveRDS(df, "./Data/feature_set.rds")

df = df[df$l2fc < 0 | df$l2fc > 2.8, ]
df$Loc = "Nuclear"
df$Loc[df$l2fc < 0] = "Cytosol"
df$Loc = as.factor(df$Loc)
dim(df)
table(df$Loc)


set.seed(54321)
spec = c(train = .7, test = .15, validate = .15)

g = sample(cut(
  seq(nrow(df)), 
  nrow(df)*cumsum(c(0,spec)),
  labels = names(spec)
))

res = split(df, g)

lapply(res, dim)
#Check splits dont contain duplicates == FALSE
table(duplicated(c(rownames(res$train), rownames(res$test), rownames(res$validate))))

# Save final dataframes
saveRDS(res, "./Data/Training_frames.rds")
```

## Deep Learning

```{r, DNN, eval = FALSE}
##Run on cluster
res = readRDS("./Data/Training_frames.rds")
library(h2o)
localH2O = h2o.init(ip="localhost", port = 54321, startH2O = TRUE, nthreads = 24, max_mem_size = "48G")
train_hex = as.h2o(res$train[ ,-1])
valid_hex = as.h2o(res$validate[ ,-1])
test_hex = as.h2o(res$validate[ ,-1])

hyper_params <- list(
  activation=c("RectifierWithDropout"),
  hidden =list( c(500,250,125), c(256, 128,64) c(128,64,32), c(64,32,16), ),
                input_dropout_ratio=c(0, 0.1, 0.2, 0.3, 0.4),
                l1 = c(1e-5, 1e-4, 1e-3, 0),
                l2 = c(1e-5, 1e-4, 1e-3, 0)
  )
  
search_criteria = list(strategy = "RandomDiscrete", max_models = 800, seed=54321, stopping_rounds=5, stopping_tolerance=1e-3)
  
  dl_random_grid <- h2o.grid(
    algorithm="deeplearning",
    grid_id = "dl_grid",
    training_frame=train_hex,
    validation_frame = valid_hex,
    x=1:1582,
    y=1583,
    epochs= 400,
    stopping_metric="misclassification",
    stopping_tolerance=1e-2,        
    stopping_rounds=5,
    hyper_params = hyper_params,
    search_criteria = search_criteria
  )
  
grid <- h2o.getGrid("dl_grid",sort_by="accuracy", decreasing=TRUE)
  
best_model <- h2o.getModel(grid@model_ids[[1]])
best_model
 
h2o.saveModel(best_model, path="./Data/Models/", force=TRUE)
file.rename(paste0("./Data/Models/", best_model@model_id), "./Data/Models/DeepLncRNA")
h2o.shutdown(FALSE)
  
```

##Create Models for comparison

```{r, model_building}
library(randomForest)
library(doParallel)
library(caret)


res = readRDS("./Data/Training_frames.rds")
cluster <- makeCluster(detectCores()) 
registerDoParallel(cluster)

logit <- caret::train(Loc ~.,
                           data = res$train[ ,-1],
                           method = "LogitBoost",
                           preProcess = c("center", "scale"),
                           trControl = trainControl(classProbs = TRUE),
                           allowParralel = TRUE,
                           tuneLength = 6)

gbm <- caret::train(Loc ~.,
                           data = res$train[ ,-1],
                           method = "gbm",
                           preProcess = c("center", "scale"),
                           trControl = trainControl(classProbs = TRUE),
                           allowParralel = TRUE,
                           tuneLength = 6)

svm_radial <- caret::train(Loc ~.,
                           data = res$train[ ,-1],
                           method = "svmRadial",
                           preProcess = c("center", "scale"),
                           trControl = trainControl(classProbs = TRUE),
                           allowParralel = TRUE,
                           tuneLength = 6)

rf <- caret::train(Loc ~ ., data = res$train[ ,-1], method = "rf",
                           allowParralel = TRUE,
                           tuneLength = 6,
                           ntree = 101)

stopCluster(cluster)
registerDoSEQ()

saveRDS(gbm, "./Data/Models/gbm.rds")
saveRDS(logit, "./Data/Models/Logit.rds")
saveRDS(svm_radial, "./Data/Models/SVM.rds")
saveRDS(rf, "./Data/Models/RF.rds")
```


 
```{r, Feature_evaluation}


```

 ## Machine learning model comparison on validation set
 
```{r, model_comparison, fig.height= 7, fig.width= 5}
library(h2o)
library(caret)
logit = readRDS("./Data/Models/Logit.rds")
gbm = readRDS("./Data/Models/gbm.rds")
rf = readRDS("./Data/Models/RF.rds")
SVM = readRDS("./Data/Models/SVM.rds")

localH2O = h2o.init(ip="localhost", port = 54321, startH2O = TRUE, nthreads = 1, max_mem_size = "2G")
DNN = h2o.loadModel("./Data/Models/DeepLncRNA")


train_hex = as.h2o(res$train)
valid_hex = as.h2o(res$validate)
test_hex = as.h2o(res$test)


## Make plot of validation accuracies
log_vals = predict(logit, res$validate)
log_conf = confusionMatrix(log_vals, res$validate$Loc, positive = "Nuclear")

gbm_vals = predict(gbm, res$validate)
gbm_conf = confusionMatrix(gbm_vals, res$validate$Loc, positive = "Nuclear")

rf_vals = predict(rf, res$validate)
rf_conf = confusionMatrix(rf_vals, res$validate$Loc, positive = "Nuclear")

svm_vals = predict(SVM, res$validate)
svm_conf = confusionMatrix(svm_vals, res$validate$Loc, positive = "Nuclear")

DNN_vals = predict(DNN, valid_hex)
DNN_conf = confusionMatrix(as.factor(as.vector(DNN_vals$predict)), res$validate$Loc, positive = "Nuclear")



GetMets = function(confMat){
  ## get acc, sens, spec, F1 from caret confusion matrix
  acc = confMat$overall[1]
  spec =  confMat$table[1,1]/(confMat$table[1,1] + confMat$table[2,1])
  sens = confMat$table[2,2]/(confMat$table[2,2] + confMat$table[1,2])
  prec = confMat$table[2,2]/(confMat$table[2,2] + confMat$table[2,1])
  F1 = (2*prec*sens )/ (prec+sens)
  mets = c(acc,sens,spec,F1)
  names(mets) = c("Accuracy","Sensitivity","Specificity","F1")
  return(mets)
}

valid_mets = data.frame( Log = GetMets(log_conf),GBM = GetMets(gbm_conf), RF = GetMets(rf_conf), SVM = GetMets(svm_conf), DeepLncRNA = GetMets(DNN_conf), Metrics = names(GetMets(DNN_conf))) %>% 
  tidyr::gather(valid_mets, Metrics)
valid_mets$Metric = rep(c("Accuracy","Sensitivity","Specificity","F1"), 5)
colnames(valid_mets) = c("Model","Value", "Metric")

ggplot(data = valid_mets, aes(as.factor(Model), Value, fill = Metric) ) +
  geom_bar( stat = "identity", position = "dodge") +
  scale_fill_brewer(palette = "Dark2") +
  coord_cartesian(ylim=c(0,1)) +
  xlab("Model") +
  theme_bw() + 
  theme(axis.text=element_text(size=13),
  axis.title=element_text(size=14,face="bold"), legend.title = element_text(size=12,face="bold"))
ggsave("./Figures/Model_Comparison.pdf")

dnn_test = predict(DNN, test_hex)
rf_test = predict(rf, res$test, "prob")
svm_test = predict(SVM, res$test, "prob")
gbm_test = predict(gbm, res$test, "prob")
logit_test = predict(logit, res$test, "prob")
h2o.shutdown(FALSE)

```


## Model evalutation on test set

```{r, ROC, fig.height= 6, fig.width= 6}


gbm_roc =pROC::roc(predictor=as.vector(gbm_test$Nuclear),
                  response=res$test$Loc,
                  levels=levels(res$test$Loc) )

logit_roc =pROC::roc(predictor=as.vector(logit_test$Nuclear),
                  response=res$test$Loc,
                  levels=levels(res$test$Loc) )

dnn_roc = pROC::roc(predictor=as.vector(dnn_test$Nuclear),
                    response=res$test$Loc,
                    levels=levels(res$test$Loc) )


plot(dnn_roc, col = "blue", cex.lab = 1.3, cex.axis = 1, lwd=3)

rf_roc =pROC::roc(predictor=as.vector(rf_test$Nuclear),
                  response=res$test$Loc,
                  levels=levels(res$test$Loc) )

plot(rf_roc, add = TRUE, col = "darkorange",lty=2, lwd = 3)

svm_roc = pROC::roc(predictor=as.vector(svm_test$Nuclear),
                     response=res$test$Loc,
                     levels=levels(res$test$Loc) )
plot(svm_roc, add = TRUE, col = "magenta",lty =3, lwd = 3)

plot(logit_roc, add = TRUE, col = "red",lty=4, lwd = 3)
plot(gbm_roc, add = TRUE, col = "darkgreen",lty=5, lwd = 3)


legend(0.35, 0.35, legend=c("DNN", "RF", "SVM","Log","GBM", "Random Guess"),
       col=c("blue", "darkorange","magenta","red","green","grey"), lty = c(1, 2,3,4,5,1), lwd = c(3,3,3,2,2,2), cex = 1.2)
h2o.shutdown(FALSE)
```

## Test set performance metrics

```{r, testset_metrics}
rf_test = predict(rf, res$test)
svm_test = predict(SVM, res$test)
gbm_test = predict(gbm, res$test)
logit_test = predict(logit, res$test)


gbm_conf = confusionMatrix(gbm_test, res$test$Loc, positive = "Nuclear")
logit_conf = confusionMatrix(logit_test, res$test$Loc, positive = "Nuclear")
rf_conf = confusionMatrix(rf_test, res$test$Loc, positive = "Nuclear")
svm_conf = confusionMatrix(svm_test, res$test$Loc, positive = "Nuclear")
dnn_conf = confusionMatrix(as.factor(as.vector(dnn_test$predict)), res$test$Loc, positive = "Nuclear")

test_mets = data.frame(t(data.frame(GBM = GetMets(gbm_conf), Log = GetMets(log_conf), RF = GetMets(rf_conf), SVM = GetMets(svm_conf), DNN = GetMets(dnn_conf)) ))
test_mets$AUC = c(rf_roc$auc, svm_roc$auc, dnn_roc$auc, gbm_roc$auc, logit_roc$auc)

gbm_mcc = mltools::mcc(preds = as.numeric(gbm_test)-1, actuals = as.numeric(res$test$Loc)-1)
logit_mcc = mltools::mcc(preds = as.numeric(logit_test)-1, actuals = as.numeric(res$test$Loc)-1)
svm_mcc = mltools::mcc(preds = as.numeric(svm_test)-1, actuals = as.numeric(res$test$Loc)-1)
dnn_mcc = mltools::mcc(preds = as.numeric(as.factor(as.vector(dnn_test$predict)))-1, actuals = as.numeric(res$test$Loc)-1)

test_mets$MCC = c(gbm_mcc, logit_mcc, rf_mcc, svm_mcc, dnn_mcc)
write.csv(round(test_mets, 3), "./Data/test_metrics.csv")

```


## Differentially localized lncRNA heatmap

```{r, lncRNA_heatmap, fig.height= 7, fig.width= 5}
mat_match = match(rownames(mat), rownames(df))
mat = mat[!is.na(mat_match), ]

div_cols = unlist(list(color = brewer.pal(11, "RdYlBu")))[11:1]
mat2 = mat
mat2[mat2 >= 7] = 7
mat2[mat2 <= -5] = -5
mat_breaks = c(min(mat2, na.rm = TRUE), -3, -1,  0,  1, 3, 5 ,max(mat2, na.rm = TRUE))
div_cols = div_cols[c(1,2,3,7,8,10,11)]

mat2 = mat2[order(rowMeans(mat2, na.rm = TRUE), decreasing = TRUE), ]

pheatmap(mat2, cluster_rows = FALSE , cluster_cols = TRUE, labels_row = "", breaks = mat_breaks, color = div_cols, legend_breaks = c(-3, -1,0, 1, 3,5, max(mat2, na.rm = TRUE)), legend_labels = c(-3, -1, 0, 1, 3,5, "L2FC\n"), cex = 1.2 ) 

```


## Genomic predictions

```{r, genomic_analysis, warning = FALSE}
Gtrans = GetSeq(annots$transcript_id, mart)
Gtrans = Gtrans[is.na(match(Gtrans$ensembl_transcript_id, rownames(res$train))), ]
Gdf = FeatureExtract(Gtrans)
preds = DeepLncRNA(Gdf)    
Gdf = cbind(preds, Gdf)
Gdf$transcript_biotype = Gtrans$transcript_biotype
dim(Gdf)

## check overall predictions
table(preds$predict)

 
pos = Gdf %>%
  group_by(transcript_biotype) %>%
  summarise(avg = (sum(predict == "Nuclear")/n())*100, count = n()) %>%
  filter(count > 20) %>%
  arrange(desc(avg)) %>%
  dplyr::select(transcript_biotype)
pos = as.character(pos$transcript_biotype)

Gdf %>% 
  group_by(transcript_biotype) %>% 
  summarise(avg = (sum(predict == "Nuclear")/n())*100, count = n()) %>% 
  filter(count > 20) %>% 
  arrange(desc(avg)) %>% 
  ggplot( aes(x = as.factor(transcript_biotype), avg, fill = as.factor(transcript_biotype))) +
  geom_bar(stat="identity", position = "dodge") +
  xlab("") +
  ylab("% Nuclear") +
  coord_flip() +
  geom_hline(yintercept = 50, col = "red", lwd = 1) +
  scale_x_discrete(limits = pos) +
  guides(fill=FALSE) +
  theme_bw(base_size = 18)
ggsave("./Figures/genomic_biotype_loc.pdf")

```

## Predictions on lncRNAs with known  localizations

```{r, known_locs}
library(dplyr)
library(reshape2)
library(ggplot2)
# annots ------------------------------------------------------------------
known = read.csv("./Data/annotated_lncRNAs.csv")
source("./src/DeepLncRNA.R")
trans = Gene2Transcript(known$ID, mart)
seqs = GetSeq(trans$ensembl_transcript_id, mart)

mousemart = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
transM = Gene2Transcript(known$ID[known$Species == "Mouse"], mart = mousemart)
seqM = GetSeq(transM$ensembl_transcript_id, mart = mousemart)

Borg = seqM
Borg$ensembl_transcript_id = "BORG"
Borg$cdna = stringr::str_replace_all(known$seq[6], "\n","")
Borg$transcript_length = nchar(as.character(Borg$cdna))
Borg$GC_content = sum(oligonucleotideFrequency(DNAString(Borg$cdna), 1)[c(2,3)])/ sum(oligonucleotideFrequency(DNAString(Borg$cdna), 1))


fullseq = rbind(Borg,seqM,seqs)
fullseq$chromosome_name[1:2] = 0 ## set to zero bc not on human chroms (mouse lncRNAs)
fullseq = fullseq %>% filter(transcript_biotype != "protein_coding") %>% 
  filter(transcript_biotype != "nonsense_mediated_decay") %>% 
  filter(transcript_length >= 200)

df = FeatureExtract(fullseq)

trans = rbind(transM, trans, c("BORG","BORG","lincRNA"))

################
#build new DNN excluding all known LncRNAs from train/valid set
 res$train = res$train[is.na(match(rownames(res$train), trans$ensembl_transcript_id)), ]
res$validate = res$validate[is.na(match(rownames(res$validate), trans$ensembl_transcript_id)), ]

library(h2o)
localH2O = h2o.init(ip="localhost", port = 54321, startH2O = TRUE, nthreads = -1, max_mem_size = "15G")
h2o.no_progress()
train_hex = as.h2o(res$train[ ,-1])
valid_hex = as.h2o(res$validate[ ,-1])

DNN = h2o.deeplearning(
    training_frame=train_hex,
    validation_frame = valid_hex,
    x=1:1582,
    y=1583,
    epochs= 2000,
    stopping_metric="misclassification",
    stopping_tolerance=1e-3,        
    stopping_rounds=10,
    activation = "RectifierWithDropout",
    hidden = c(64,32,16),
    input_dropout_ratio = 0.3,
    l1 = 1e-3,
    l2 = 1e-5)

preds = predict(DNN, as.h2o(df))
preds = as.data.frame(preds)
rownames(preds) = df$ensembl_transcript_id

preds$gene = trans$ensembl_gene_id[match(rownames(preds), trans$ensembl_transcript_id)]
preds$name = known$Name[match(preds$gene, known$ID)]
preds$Loc = known$Localized[match(preds$gene, known$ID)]

meds = preds %>% 
  group_by(gene) %>% 
  summarise(Nuclear = sum(predict == "Nuclear")/n()) 

meds$Loc = known$Localized[match(meds$gene, known$ID)]
meds = meds[order(meds$Loc), ]
meds$Cytosol = 1-meds$Nuclear
nams=  known$Name[match( meds$gene, known$ID)]
nams = nams[!is.na(nams)]
meds$Name = nams

medL = melt(meds[,c(2,4,5)], id.vars = "Name")

colvec = rep("red", nrow(meds))
colvec[meds$Loc == "Cytosol"] = "blue"
colvec[meds$Loc == "Dual"] = "black"

ggplot(medL, aes(x = Name, value, fill = as.factor(variable))) +
  geom_bar(stat="identity") +
  xlab("") +
  ylab("Percent Localized") +
  coord_flip() +
  scale_x_discrete(limits = meds$Name) +
  scale_fill_discrete(name = "Fraction") +
  theme_bw() +
  theme(axis.text.y=element_text(colour = colvec, size = 12))
ggsave("./Figures/annotated_lncRNAs.pdf", width = 5, height = 7, units = "in")

```





